src/lib/hooks/swap/useSwapApproval.ts

```diff
@@ -155,51 +160,52 @@ export const useApproveAndPermit = (
 ) => {
   const deadline = useTransactionDeadline()
 
-  // Check approvals on ERC20 contract based on amount
+  // Check approvals on ERC20 contract based on amount.
   const [approval, getApproval] = useSwapApproval(trade, allowedSlippage, useIsPendingApproval, amount)
 
-  // Check status of permit and wether token supports it
+  // Check status of permit and whether token supports it.
   const {
     state: signatureState,
     signatureData,
     gatherPermitSignature,
   } = useERC20PermitFromTrade(trade, allowedSlippage, deadline)
 
-  const notApproved =
-    approval === ApprovalState.NOT_APPROVED &&
-    !(signatureState === UseERC20PermitState.SIGNED && Boolean(gatherPermitSignature))
+  const notApproved = approval === ApprovalState.NOT_APPROVED && !(signatureState === UseERC20PermitState.SIGNED)
 
-  // setup callback if need to submit approval on chain
-  const addTransaction = useAddTransaction()
-  const addApprovalTransaction = useCallback(() => {
-    getApproval().then((transaction) => {
-      if (transaction) {
-        addTransaction({ type: TransactionType.APPROVAL, ...transaction })
-      }
-    })
-  }, [addTransaction, getApproval])
-
-  // if permit is supported, trigger a signature, if not create approval transaction
+  // If permit is supported, trigger a signature, if not create approval transaction.
   const handleApproveOrPermit = useCallback(async () => {
     if (signatureState === UseERC20PermitState.NOT_SIGNED && gatherPermitSignature) {
       try {
-        await gatherPermitSignature()
+        return await gatherPermitSignature()
       } catch (error) {
-        // try to approve if gatherPermitSignature failed for any reason other than the user rejecting it
+        // Try to approve if gatherPermitSignature failed for any reason other than the user rejecting it
         if (error?.code !== 4001) {
-          addApprovalTransaction()
+          return getApproval()
         }
       }
     } else {
-      addApprovalTransaction()
+      return getApproval()
+    }
+  }, [signatureState, gatherPermitSignature, getApproval])
+
+  const approvalState = useMemo(() => {
+    if (approval === ApprovalState.PENDING) {
+      return ApproveOrPermitState.PENDING_APPROVAL
+    }
+    if (signatureState === UseERC20PermitState.LOADING) {
+      return ApproveOrPermitState.PENDING_SIGNATURE
+    }
+    if (notApproved && Boolean(gatherPermitSignature)) {
+      return ApproveOrPermitState.REQUIRES_SIGNATURE
+    }
+    if (notApproved) {
+      return ApproveOrPermitState.REQUIRES_APPROVAL
     }
-  }, [signatureState, gatherPermitSignature, addApprovalTransaction])
+    return undefined
```

Comment:
Should this be another state, ApprovedOrPermitState.APPROVED?

Reply:
yeah i like that